### DISCUSSION TITLE
Collapsing <head> Behavior?

### DISCUSSION THREAD

**Discussion by FredKSchott** [withastro/roadmap#15] [Proposal]
@jonathantneal and I were chatting about #12 and discussing `<head>` behavior & bugs in Astro, and how Svelte solves this via `<svelte:head>`. We came to the realization that only supporting `<head>` in your top-level pages (and not in components) doesn't make sense for anyone using a common layout component, which can act as a component but often defines the entire `<head>`. This is a very common use-case for Astro users, especially as we plan to lean more into common layouts as a way to do Astro themes: https://docs.astro.build/core-concepts/layouts/

An idea that we came to was to just always collapse the `<head>` values of any components in the tree into the final page `<head>` of the rendered HTML. This would solve for our layout use-case, and also give us more flexible `head` injection inside of components at the same time. 

The main reasoning why this "magic" feels okay is that `<head>` has no valid use-case (that we are aware of) inside a component template, since nesting a `<head>` inside of a `<body>` would be invalid.

**Is there buy-in for this idea, conceptually? I can flesh this out into a full RFC if so, but I wanted to do a quick temperature check before writing more.**

Credit to @jonathantneal who (iirc) first proposed this in our convo!

**Comment by drwpow** [comment]
+1 to the overall concept. One potential pain point that I’d like to see expanded upon would be what people have been experiencing in CSS: expected load order. People have different assumptions about the order in which CSS is collected when imported across multiple pages & components. Merging `<head>`s would have the same concerns with the addition of `<script>`s (e.g. which takes priority—the page? or the last-imported component?). I’d like to see details on our desired ordering when `<head>` tags are merged. And even if there’s not a unanimous buy-in, we at least have some predictable behavior that we enforce.

**Comment by jonathantneal** [comment]
Overall, I was hoping that all `.astro` files would be treated like fragments. And then a page rendering would be 'special' in that it would collapse any `<head>` trees into the first one.

**Comment by tony-sull** [comment]
I'm a big fan of how the general idea of something similar to `<svelte:head>`.  It's particularly helpful for SEO meta tags, a page layout can set all default all the required SEO tags then the page, sub-navigation layout, or components on the page can all override with more specific data

One big question would be if priority is given based on a depth-first or breadth-first search.  i.e. What's the title on this page?

```html
<body>
    <div>
        <div>
            <head>
                <title>Other title</title>
            </head>
        <div>
    <div>
    <head>
        <title>Best title ever</title>
    </head>
</body>
```

**Reply by FredKSchott** [reply]
alternatively, could we warn if we saw things like this?

But to the general point, I think that's a great callout. I'd be fine with whichever was easier to implement, as long as it was explicit and not changing over time.

**Reply by jonathantneal** [reply]
It would hope it 'works like the web', which means parsing order (I think). In the example, this would appear as tho it were depth first. This can be observed in a browser with the example code.

Based on my understanding of the proposal, in the example I would expect something like this:

- When `<body` is parsed:
  - This would initialize implicit **html**, **head**, and **body** fragments, which would not yet render as tags.
  - The **body** fragment would assign `body` to its tag name, making the fragment 'visible'.
  - The contents of this element would be appended to the **body** as parsing continues.
- When the first `<head` is parsed:
  - The **head** fragment would assign `head` to its tag name, making it 'visible'.
  - The contents of this element would be appended to the **head** as parsing continues.
  - ... until the first `</head` is parsed, at which point appending goes back to **body**.
- When the second `<head` is parsed:
  - The **head** would assign `head` to its tag name; making no change.
  - The contents of this element would be appended to the **head** as parsing continues.
  - ... until the second `</head` is parsed, at which point appending goes back to **body**.

**Reply by tony-sull** [reply]
> It would hope it 'works like the web'

All for matching existing browser behavior here!  This is one of those corner cases that I never knew if there was a spec for (and apparently never tried), or if browsers ever bothered to make sure they actually match the spec :smile:

@FredKSchott I could see that being a really useful warning behind a verbose flag or similar, I'd just be hesitant to always log the warning since there are reasons why it would be done intentially (it's pretty common to override `<title>` in Sapper/SvelteKit apps multiple times if you use nested layouts)

**Reply by FredKSchott** [reply]
That's a good point, I like the idea of overriding key head elements for that use-case. As a feature, we would just need to be aware of which elements to overwrite (title, meta tags by name, etc).

I think that can be considered a separate feature request from this proposal, but I'm also not opposed to combining with this RFC if there's interest and that idea is not controversial.

**Reply by FredKSchott** [reply]
@jonathantneal check out the RFC, would love your thoughts! I like how detailed your comment is on the parse order behavior, maybe we could bring some of that into the RFC.

FYI I've shifted a bit from when we last talked on the topic of _outputting_ HTML without a `head` `body` or `html` tag. Still +1 on supporting authoring this way, but in terms of HTML output I'm less convinced of its value. Check out the alternatives section of the RFC for more detail on that.

**Reply by FredKSchott** [reply]
https://github.com/withastro/rfcs/discussions/15#discussioncomment-1726197

Now that there's an RFC, it would be great to move this convo into that PR so that the feedback here isn't lost.

**Comment by FredKSchott** [comment]
Turned into an RFC! https://github.com/withastro/rfcs/pull/26

**Comment by florian-lefebvre** [comment]
RFC was merged
